compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5606,13 +5571,86 @@ private BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType c
         }
     }
 
-    private BType getQueryTableType(BLangQueryExpr queryExpr, BType constraintType) {
+    void solveSelectTypeAndResolveType(BLangQueryExpr queryExpr, BLangExpression selectExp, BType type,
+                                       BType collectionType, List<BType> selectTypes, List<BType> resolvedTypes,
+                                       SymbolEnv env, AnalyzerData data, boolean isReadonly) {
+        BType selectType, resolvedType;
+        type = Types.getReferredType(type);
+        switch (type.tag) {
+            case TypeTags.ARRAY:
+                BType elementType = ((BArrayType) type).eType;
+                if (elementType.tag == TypeTags.INTERSECTION) {
+                    selectType = checkExpr(selectExp, env, ((BIntersectionType) elementType).effectiveType, data);
+                } else {
+                    selectType = checkExpr(selectExp, env, elementType, data);
+                }
+                resolvedType = new BArrayType(selectType);
+                if (resolvedType.tag != TypeTags.SEMANTIC_ERROR && (isReadonly ||
+                        Symbols.isFlagOn(type.flags, Flags.READONLY))) {
+                    resolvedType = ImmutableTypeCloner.getImmutableIntersectionType(null, types, resolvedType, env,
+                            symTable, anonymousModelHelper, names, null);
+                }
+                break;
+            case TypeTags.TABLE:
+                selectType = checkExpr(selectExp, env, types.getSafeType(((BTableType) type).constraint,
+                        true, true), data);
+                resolvedType = symTable.tableType;
+                if (resolvedType.tag != TypeTags.SEMANTIC_ERROR && (isReadonly ||
```

Comment:
I think we don't need the OR condition here. checking only `isReadOnly` with `SEMANTIC_ERROR` would suffice.

Reply:
We need to check ReadOnly flag too because, `isReadOnly` flag will not be triggered for some readonly scenarios. 

Ex.
```ballerina
type T readonly & record {
   string[] params;
};

function foo(string[] v) returns T {
  return { params: from var s in v select s };
}
```